### ISSUE TITLE
Efficient batchwise dropout training using submatrices

### ISSUE THREAD

**Issue by robertsdionne** [BVLC/caffe#2701]
Generalize classes InnerProductLayer and ConvolutionLayer to support sub-matrix multiplication to more efficiently implement dropout and to use a single dropout mask per minibatch rather than per example.

Efficient batchwise dropout training using submatrices
Authors: Ben Graham, Jeremy Reizenstein, Leigh Robinson
[Abstract](http://arxiv.org/abs/1502.02478)
[Full text](http://arxiv.org/pdf/1502.02478v1.pdf)

Using the above technique we can realize 42% savings in computation time using mini-batchwise dropout over independent dropout for a 3-layer neural network with 500 neurons per hidden layer, with increasing savings as layer dimensions get larger (see figure 1).

Notable quotes:
- "Consider a network with three hidden layers of size 1000, trained for 1000 epochs using minibatches of size 100. The number of distinct dropout patterns, 2^3784, is so large that we can assume that we will never generate the same dropout mask twice."
- With batchwise dropout, "As the number of dropout masks increases into the hundreds, we see that it is quickly a case of diminishing returns."
- "Using independent dropout with recurrent neural networks can be too disruptive
  to allow effective learning; one solution is to only apply dropout to some parts
  of the network [12]. Batchwise dropout may provide a less damaging form of
  dropout, as each unit will either be on or off for the whole time period."

**Comment by shelhamer** [comment]
Closing for lack of interest—regular dropout is a trivial amount of computation and seems to work well enough (with the exception of training recurrent nets). Thanks for sharing the reference all the same.
